phoenix-core-server/src/main/java/org/apache/phoenix/replication/ReplicationLogWriter.java

```diff
@@ -0,0 +1,400 @@
+        "phoenix.replication.writer.sync.retry.pause.ms";
+    public static final long DEFAULT_REPLICATION_WRITER_SYNC_RETRY_PAUSE = 100;
+
+    private static final byte EVENT_TYPE_DATA = 0;
+    private static final byte EVENT_TYPE_SYNC = 1;
+
+    private final Configuration conf;
+    private final ReplicationLogManager logManager;
+    private final int ringBufferSize;
+    private final long syncTimeoutMs;
+    private Disruptor<LogEvent> disruptor;
+    private RingBuffer<LogEvent> ringBuffer;
+    private final AtomicBoolean started = new AtomicBoolean(false);
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    public ReplicationLogWriter(Configuration conf, ReplicationLogManager logManager) {
+        this.conf = conf;
+        this.logManager = logManager;
+        this.ringBufferSize = conf.getInt(REPLICATION_WRITER_RINGBUFFER_SIZE_KEY,
+            DEFAULT_REPLICATION_WRITER_RINGBUFFER_SIZE);
+        this.syncTimeoutMs = conf.getLong(REPLICATION_WRITER_SYNC_TIMEOUT_KEY,
+            DEFAULT_REPLICATION_WRITER_SYNC_TIMEOUT);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void init() {
+        // Initialize the Disruptor. We use ProducerType.MULTI because multiple handlers might
+        // call append concurrently. We use YieldingWaitStrategy for low latency. When the ring
+        // buffer is full (controlled by REPLICATION_WRITER_RINGBUFFER_SIZE_KEY), producers
+        // calling ringBuffer.next() will effectively block (by yielding/spinning), creating
+        // backpressure on the callers. This ensures appends don't proceed until there is space.
+        disruptor = new Disruptor<>(LogEvent.EVENT_FACTORY, ringBufferSize,
+            new ThreadFactoryBuilder().setNameFormat("ReplicationLogEventHandler-%d")
+                .setDaemon(true).build(),
+            ProducerType.MULTI, new YieldingWaitStrategy());
+        disruptor.handleEventsWith(new LogEventHandler(conf, logManager));
+        disruptor.setDefaultExceptionHandler(new LogExceptionHandler());
+        ringBuffer = disruptor.start();
+        started.set(true);
+        LOG.info("ReplicationLogWriter started with ring buffer size {}", ringBufferSize);
+    }
+
+    /**
+     * Append a mutation to the log. This method is non-blocking and returns quickly, unless the
+     * ring buffer is full. The actual write happens asynchronously. We expect multiple append()
+     * calls followed by a sync(). The appends will be batched by the Disruptor. Should the ring
+     * buffer become full, which is not expected under normal operation but could (and should)
+     * happen if the log file writer is unable to make progress, due to a HDFS level disruption.
+     * Should we enter that condition this method will block until the append can be inserted.
+     * @param tableName The name of the HBase table the mutation applies to.
+     * @param commitId  The commit identifier (e.g., SCN) associated with the mutation.
```

Comment:
Is this basically the timestamp the coproc assigns to the mutation ?

Reply:
No. The mutation and the cells in the mutation carry that timestamp. The "commit identifier" is a Phoenix level transaction id. It can be whatever Phoenix needs it to be. If we don't need this concept at this layer it can be removed.